### PR TITLE
[feature](nereids) support subquery combine

### PR DESCRIPTION
**Pull request by xzj7019**
# Proposed changes

Support subquery combination optimization. 

## Problem summary

This pr supports subquery combination optimization. This kind of combination can help to identify and eliminate similar subqueries which have containment relationship and have a performance gain.

Currently only support exist-notexists combination type, i.e, if we find a exists subquery is contained by a notexists subquery, the not exists one can be eliminated and move the excepted expressions, such as filter conditions into exists subquery with additional having sum(case when excepted_filters then 1 else 0) = 0 constraint.

Describe your changes.

This pr adds a subquery combine rule in analyze stage, since the subquery removing action should be identified before SubqueryToApply. This rule will check all the preconditions the optimization will be taken. 

A UnCorrelatedApplyHavingProjectFilter rewrite rule will be taken after subquery transformed, and a special having expression will be set into apply and finally into semi-join with a similar way. At last, a general designed semi-to-inner rewrite rule is added to identify this having expression, which marked this semi-to-inner will use inner + gby pattern, and the having expression will follow this generated gby. 

Note that subquery combination is designed for general purpose but currently only support exists-notexists, the rest types, such as exists-exists, etc, will be supported in the future. So as the semi-to-inner rule, currently only support inner + gby pattern under some special cases, but the rest common types, such as pure inner and gby inner pattern will be supported in the future. 

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/ExpressionUtils.java [hunk 1]
```diff
@@ -111,6 +117,16 @@ private static void extract(Class<? extends Expression> type, Expression expr, C
...
+    private static void extract(Expression expr, Set<SlotReference> allSlotExprs) {
+        if (expr instanceof SlotReference) {
+            allSlotExprs.add((SlotReference) expr);
+        } else {
+            for (Expression child : expr.children()) {
+                extract(child, allSlotExprs);
+            }
+        }
+    }
+
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/ExpressionUtils.java:129]: If we want to get all slots used by expr.

```java
expr.getInputSlots()
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/ExpressionUtils.java [hunk 2]
```diff
@@ -75,6 +75,12 @@ public static Set<Expression> extractConjunctionToSet(Expression expr) {
         return exprSet;
     }
 
+    public static Set<SlotReference> extractSlotToSet(Expression expr) {
+        Set<SlotReference> exprSet = Sets.newHashSet();
+        extract(expr, exprSet);
+        return exprSet;
+    }
+
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/ExpressionUtils.java:83]: use expr.getInputSlots()

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/NereidsRewriter.java [hunk 3]
```diff
@@ -158,6 +159,11 @@ public class NereidsRewriter extends BatchRewriteJob {
             ),
 
             topic("Rewrite join",
+                // transform semi join to inner join
+                // TODO: current rule is under special scenario, will put it into cost based
+                // transformation framework in the future.
+                custom(RuleType.SEMI_TO_INNER, SemiToInner::new),
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/NereidsRewriter.java:165]: Don't mix two feature in one PR

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/UnCorrelatedApplyHavingProjectFilter.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ * Input(output:b)         Project(output:a)
+ *                           |
+ *                         Filter(UnCorrelated predicate)
+ *                           |
+ *                          child
+ */
+public class UnCorrelatedApplyHavingProjectFilter extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalApply(any(), logicalHaving(logicalProject(logicalFilter())))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/UnCorrelatedApplyHavingProjectFilter.java:67]: add an annotation `@DependsRules({SubqueryCombine.class})`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        GBY_INNER,
+        INNER_GBY,
+        UNKNOWN
+    }
+
+    @Override
+    public Plan rewriteRoot(Plan plan, JobContext jobContext) {
+        return plan.accept(this, new SemiToInnerContext(false, TransformPattern.UNKNOWN,
+                                                                new HashSet<>(), new ArrayList<>(),
+                                        Optional.empty()));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java:73]: the code is not easy for read, we could use a constructor with no paramters and init all attribute in itself

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        // only support inner gby pattern currently.
+        if (join.getSubQueryCombinedHavingExpr().isPresent()) {
+            return TransformPattern.INNER_GBY;
+        } else {
+            return TransformPattern.UNKNOWN;
+        }
+    }
+
+    private LogicalPlan transformToInnerGby(LogicalPlan plan, SemiToInnerContext context) {
+        if (!(plan instanceof LogicalProject || plan instanceof LogicalFilter)) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java:149]:
```java
if (plan instanceof LogicalProject || plan instanceof LogicalFilter) {
    return plan;
}
...
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        }
+    }
+
+    private LogicalPlan transformToInner(LogicalPlan plan, SemiToInnerContext context) {
+        return plan;
+    }
+
+    private LogicalPlan transformToGbyInner(LogicalPlan plan, SemiToInnerContext context) {
+        return plan;
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java:201]: add TODO

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Transform semi join to inner join.
+ * TODO: only support semi -> inner + gby pattern currently
+ * Will support pure inner and gby + inner patterns.
+ * Will put into cost based transformation framework.
+ */
+public class SemiToInner extends DefaultPlanRewriter<SemiToInnerContext> implements CustomRewriter {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java:57]: add an annotation `@DependsRules({SubQeueryCombine.class, UncorrelatedApplyHavingProjectFilter.class})`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+            Set<Expression> havingExprAliasList = new HashSet<>();
+            // having filter use alias = 0
+            EqualTo havingExpr = new EqualTo(aggrExprAlias.toSlot(), new BigIntLiteral(0));
+            havingExprAliasList.add(havingExpr);
+
+            LogicalAggregate newAggr = new LogicalAggregate(new ArrayList<>(combinedGbyExprs),
+                    new ArrayList<>(outputExprsAfterGby), project);
+            LogicalFilter havingFilter = new LogicalFilter(havingExprAliasList, newAggr);
+            LogicalProject newProject = new LogicalProject(new ArrayList<>(combinedGbyExprs), havingFilter);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java:186]:
```suggestion
            LogicalProject newProject = new LogicalProject(ImmutableList.copyOf(combinedGbyExprs), havingFilter);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+            outputExprsAfterGby.addAll(combinedGbyExprs);
+            outputExprsAfterGby.add(aggrExprAlias);
+
+            Set<Expression> havingExprAliasList = new HashSet<>();
+            // having filter use alias = 0
+            EqualTo havingExpr = new EqualTo(aggrExprAlias.toSlot(), new BigIntLiteral(0));
+            havingExprAliasList.add(havingExpr);
+
+            LogicalAggregate newAggr = new LogicalAggregate(new ArrayList<>(combinedGbyExprs),
+                    new ArrayList<>(outputExprsAfterGby), project);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java:184]:
```suggestion
            LogicalAggregate newAggr = new LogicalAggregate(ImmutableList.copyOf(combinedGbyExprs),
                    ImmutableList.copyOf(outputExprsAfterGby), project);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+            Expression aggrExprFromHavingExpr = extractAggrExprFromHavingExpr(context.subQueryCombinedHavingExpr.get());
+            Alias aggrExprAlias = new Alias(aggrExprFromHavingExpr, aggrExprFromHavingExpr.toSql());
+            // group by output use alias expr
+            outputExprsAfterGby.addAll(combinedGbyExprs);
+            outputExprsAfterGby.add(aggrExprAlias);
+
+            Set<Expression> havingExprAliasList = new HashSet<>();
+            // having filter use alias = 0
+            EqualTo havingExpr = new EqualTo(aggrExprAlias.toSlot(), new BigIntLiteral(0));
+            havingExprAliasList.add(havingExpr);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SemiToInner.java:181]:
```suggestion
            // having filter use alias = 0
            EqualTo havingExpr = new EqualTo(aggrExprAlias.toSlot(), new BigIntLiteral(0));
            Set<Expression> havingExprAliasList = ImmutableSet.of(havingExpr);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/SubqueryExpr.java [hunk 12]
```diff
@@ -19,49 +19,95 @@
...
     }
 
     public List<Slot> getCorrelateSlots() {
         return correlateSlots;
     }
 
+    public List<Expression> getCorrelateExpressions() {
+        List<Expression> expressionList = new ArrayList<>();
+        expressionList.addAll(correlateSlots);
+        return expressionList;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/SubqueryExpr.java:80]: why do copy?

**Review by morrySnow — COMMENTED** [review]
please add ut and regression test. and add a example in commit msg

**Review by morrySnow — COMMENTED** [review]
please add ut and regression test. and add a example in commit msg

**Pull request closed without merging**